Portable Account (#59)
fairDataSociety/FIPs: text/0059-portable-account.md

- FIP: 59
- title: Portable Account
- author: Viktor Levente Tóth, @nugaon
- status: draft
- created: 2022-10-04

# Summary

> Addressing web3 login workflow where users can retreive their wallet in a decentralized manner from anywhere by providing a _username_ and a _password_.
> 

Reviewer: ```suggestion
Addressing web3 login workflow where users can retrieve their wallet in a decentralised manner from anywhere by providing a _username_ and a _password_.
```

Author: "decentralized" is a correct spelling as well.

# Context, motivation and guide level explanation

> A wallet is an assymetric keypair that represents an account in Ethereum or in Fair Data Society.

Reviewer: ```suggestion
A wallet is an asymmetric keypair that represents an account in Ethereum or in Fair Data Society.
```

The solution worked out here allows to log in to the user account from any device in the same way like in case of Web2: providing a username and a password.
Such a login can happen in a fully decentralized way using only Ethereum Name Service (ENS) and Ethereum Swarm.

This facilitates using the same wallet in different wallet handler applications (e.g. Blossom and FDP CLI), moreover, it allows to use the account from any device without the need of a pendrive or a hardware wallet.

# Reference-level explanation
The reference implementation can be found in [FDP Storage JS library](https://github.com/fairDataSociety/fdp-storage) in the `src/account` folder.

The portable account creation happens in 3 steps:
- Wallet generation
- ENS registration (username->public key)
- portable wallet (wallet seed) upload to Ethereum Swarm

The steps are detailed in the following figure

![FDS Account](../resources/fds-account.png)

## Calculating the location of the portable account
The ENS is used to map usernames to public keys. From the public key, the account's ethereum address can be derived easily for everyone.

> The Ethereum address is needed to contruct Single Owner Chunk (SOC) address along with a consensual 32 bytes of Topic.

Reviewer: ```suggestion
The Ethereum address is needed to construct Single Owner Chunk (SOC) address along with a consensual 32 bytes of Topic.
```
